Certain Native image artifacts are not copied to run image

I am using the builder-jammy-base buildpack with the spring-boot-maven-plugin to build a native image. After some troubleshooting with the Graal team (https://github.com/oracle/graal/issues/8273) it seems that certain required artifacts (jdk_library and jdk_liberary_shim .so files) are not being included in the workspace/ directory of containers using the run image generated by this buildpack. We do see them listed as "produced artifacts" in the build log, however.

## Expected Behavior
The following .so files (in bold) from the log snippet below should be included in the workspace/ directory alongside the native image executable (org.example.App):

INFO]     [creator]     Produced artifacts:
[INFO]     [creator]      /layers/paketo-buildpacks_native-image/native-image/**libawt.so** (jdk_library)
[INFO]     [creator]      /layers/paketo-buildpacks_native-image/native-image/**libawt_headless.so** (jdk_library)
[INFO]     [creator]      /layers/paketo-buildpacks_native-image/native-image/**libawt_xawt.so** (jdk_library)
[INFO]     [creator]      /layers/paketo-buildpacks_native-image/native-image/**libfontmanager.so** (jdk_library)
[INFO]     [creator]      /layers/paketo-buildpacks_native-image/native-image/**libfreetype.so** (jdk_library)
[INFO]     [creator]      /layers/paketo-buildpacks_native-image/native-image/**libjava.so** (jdk_library_shim)
[INFO]     [creator]      /layers/paketo-buildpacks_native-image/native-image/**libjavajpeg.so** (jdk_library)
[INFO]     [creator]      /layers/paketo-buildpacks_native-image/native-image/**libjvm.so** (jdk_library_shim)
[INFO]     [creator]      /layers/paketo-buildpacks_native-image/native-image/**liblcms.so** (jdk_library)
[INFO]     [creator]      /layers/paketo-buildpacks_native-image/native-image/org.example.App (executable)

## Current Behavior
The files are not included, the directory in the run container only contains the native executable:
![image](https://github.com/paketo-buildpacks/builder-jammy-base/assets/11561288/d4da88e0-64bf-4e17-be52-0986194a13c1)


## Steps to Reproduce
<!--- Provide a link to an reproduction/test case, or an unambiguous set of steps to reproduce this bug. Include notes about the software versions you're using, the environment in which you're running, code snippets to reproduce, log output, screenshots, etc.. -->
Repro:
1. Please check out the following project for a small, single-class repro of this error: [reproduce-awt-lib-error-graal](https://github.com/tabiStein/reproduce-awt-lib-error-graal)
2. Run mvn clean install in the root directory of the project to build the jar
3. Run mvn -Pnative package to build the native image, called test/repro-awt-lib-error
4. Read the log output and verify that the jdk library artifacts were produced. Attached is a copy of mine for your convenience: 
[packageOutput.txt](https://github.com/paketo-buildpacks/builder-jammy-base/files/14098062/packageOutput.txt)
5. Run docker run test/repro-awt-lib-error
6. The run fails because the libawt.so file is not in the same directory as the image executable (according to the Graal team: https://github.com/oracle/graal/issues/8273#issuecomment-1916571083)

Environment:
GraalVM version 23.0.2 (as determined by buildpack)
JDK major version: 17
OS: MacOS Sonoma 14.2.1
Architecture: 6-Core Intel Core i9
Maven version: 3.9.6
Docker version: 24.0.7

## Motivations
My company is trying to convert one of our microservices to a native executable, using your buildpack. The service uses the Apache PDDocument class, which relies on the Java awt library. Since the library is not available to the native image at runtime, the executable crashes.

Oops ! Sorry for confusion.

I have tried your maven configuration, here is the error during the maven build process

```INFO] --- spring-boot:3.3.1:build-image-no-fork (default) @ flhacker ---
[INFO] Building image 'docker.io/library/flhacker:0.0.1-SNAPSHOT'
[INFO] 
[INFO]  > Pulling builder image 'docker.io/paketobuildpacks/builder-jammy-buildpackless-tiny:latest' 100%
[INFO]  > Pulled builder image 'paketobuildpacks/builder-jammy-buildpackless-tiny@sha256:98a3c52134815d3745dde5e7626cb919739fe7f757443678f967c3246c9f45f0'
[INFO]  > Pulling run image 'docker.io/paketobuildpacks/run-jammy-tiny:latest' 100%
[INFO]  > Pulled run image 'paketobuildpacks/run-jammy-tiny@sha256:703d6c7b862e3056afd1e4dc29b59039082da337d9d862137186049e2a36e3ee'
[INFO]  > Pulling buildpack image 'docker.io/paketobuildpacks/oracle:latest' 100%
[INFO]  > Pulled buildpack image 'paketobuildpacks/oracle@sha256:5efc87d386bd5c57137cdf33ae1349236752f1311840d9ccd8a76b56c4cb2fb1'
[INFO]  > Pulling buildpack image 'docker.io/paketobuildpacks/ca-certificates:latest' 100%
[INFO]  > Pulled buildpack image 'paketobuildpacks/ca-certificates@sha256:5d3e8eb9c1dcb199e3143f26e64d06823af41ff3247fb95e00f8aea30d0efc47'
[INFO]  > Pulling buildpack image 'docker.io/paketobuildpacks/syft:latest' 100%
[INFO]  > Pulled buildpack image 'paketobuildpacks/syft@sha256:fe21acdc764f8b99219597e1ba6d81431c04cd33ad8cd7edc376ec663c1001b2'
[INFO]  > Pulling buildpack image 'docker.io/paketobuildpacks/executable-jar:latest' 100%
[INFO]  > Pulled buildpack image 'paketobuildpacks/executable-jar@sha256:43e68b6cc3d004812d4125fd7bd5189a24d9bb0df8e7e2a2e1980a57964f7144'
[INFO]  > Pulling buildpack image 'docker.io/paketobuildpacks/spring-boot:latest' 100%
[INFO]  > Pulled buildpack image 'paketobuildpacks/spring-boot@sha256:1549798611b894c9881581e5bf4e3b3f773fd72ff21c83fab787e5d80daf9fa5'
[INFO]  > Pulling buildpack image 'docker.io/paketo-buildpacks/native-image:latest' 100%
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  21.884 s
[INFO] Finished at: 2024-07-14T22:23:35+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.springframework.boot:spring-boot-maven-plugin:3.3.1:build-image-no-fork (default) on project flhacker: Execution default of goal org.springframework.boot:spring-boot-maven-plugin:3.3.1:build-image-no-fork failed: Invalid buildpack reference 'docker.io/paketo-buildpacks/native-image' -> [Help 1]
```

Looks like as standard configuration, the plugin uses
`<builder>paketobuildpacks/builder-jammy-tiny:latest</builder>`

But even with this, the error remains the same as above.

Thanks again !

So this still fails for me with
`Caused by: java.lang.UnsatisfiedLinkError: No awt in java.library.path
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.NativeLibrarySupport.loadLibraryRelative(NativeLibrarySupport.java:136)`

But :
- I see that the native image options enforce `+StaticExecutableWithDynamicLibC`. Which has **never** worked, at least for me. Bug open [here](https://github.com/oracle/graal/issues/8911#issuecomment-2118879220). Why is this enforced ? Poke @fniephaus 
- I can't, for the life of me running `bash` with `it` to check inside the container if the `.so` are there.

```
docker run -it --rm --entrypoint=bin/bash flhacker:0.0.1-SNAPSHOT 
docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "bin/bash": stat bin/bash: no such file or directory: unknown.
```

--- a/native/arguments.go
+++ b/native/arguments.go
@@ -18,7 +18,6 @@ package native
 
 import (
 	"fmt"
-	"io/ioutil"
 	"os"
 	"path/filepath"
 	"sort"
@@ -82,20 +81,20 @@ type UserFileArguments struct {
 // Configure returns the inputArgs plus the additional arguments provided via argfile, setting via the '@argfile' format
 func (u UserFileArguments) Configure(inputArgs []string) ([]string, string, error) {
 
-	rawArgs, err := ioutil.ReadFile(u.ArgumentsFile)
+	rawArgs, err := os.ReadFile(u.ArgumentsFile)
 	if err != nil {
 		return []string{}, "", fmt.Errorf("read arguments from %s\n%w", u.ArgumentsFile, err)
 	}
 
 	fileArgs := strings.Split(string(rawArgs), "\n")
-	if len(fileArgs) == 1{
+	if len(fileArgs) == 1 {
 		fileArgs = strings.Split(string(rawArgs), " ")
 	}
 
 	if containsArg("-jar", fileArgs) {
 		fileArgs = replaceJarArguments(fileArgs)
 		newArgList := strings.Join(fileArgs, " ")
-		if err = os.WriteFile(u.ArgumentsFile,[]byte(newArgList),0644); err != nil{
+		if err = os.WriteFile(u.ArgumentsFile, []byte(newArgList), 0644); err != nil {
 			return []string{}, "", fmt.Errorf("unable to write to arguments file %s\n%w", u.ArgumentsFile, err)
 		}
 	}
@@ -105,7 +104,6 @@ func (u UserFileArguments) Configure(inputArgs []string) ([]string, string, erro
 	return inputArgs, "", nil
 
 }
-
 
 // containsArg checks if needle is found in haystack
 //

--- a/native/arguments_test.go
+++ b/native/arguments_test.go
@@ -18,7 +18,6 @@ package native_test
 
 import (
 	"fmt"
-	"io/ioutil"
 	"os"
 	"path/filepath"
 	"testing"
@@ -40,13 +39,8 @@ func testArguments(t *testing.T, context spec.G, it spec.S) {
 	)
 
 	it.Before(func() {
-		var err error
-
-		ctx.Application.Path, err = ioutil.TempDir("", "native-image-application")
-		Expect(err).NotTo(HaveOccurred())
-
-		ctx.Layers.Path, err = ioutil.TempDir("", "native-image-layers")
-		Expect(err).NotTo(HaveOccurred())
+		ctx.Application.Path = t.TempDir()
+		ctx.Layers.Path = t.TempDir()
 	})
 
 	it.After(func() {
@@ -126,10 +120,10 @@ func testArguments(t *testing.T, context spec.G, it spec.S) {
 	context("user arguments from file", func() {
 		it.Before(func() {
 			Expect(os.MkdirAll(filepath.Join(ctx.Application.Path, "target"), 0755)).To(Succeed())
-			Expect(ioutil.WriteFile(filepath.Join(ctx.Application.Path, "target", "more-stuff.txt"), []byte("more stuff"), 0644)).To(Succeed())
-			Expect(ioutil.WriteFile(filepath.Join(ctx.Application.Path, "target", "more-stuff-quotes.txt"), []byte(`before -jar "more stuff.jar" after -other="my path"`), 0644)).To(Succeed())
-			Expect(ioutil.WriteFile(filepath.Join(ctx.Application.Path, "target", "more-stuff-class.txt"), []byte(`stuff -jar stuff.jar after`), 0644)).To(Succeed())
-			Expect(ioutil.WriteFile(filepath.Join(ctx.Application.Path, "target", "override.txt"), []byte(`one=output`), 0644)).To(Succeed())
+			Expect(os.WriteFile(filepath.Join(ctx.Application.Path, "target", "more-stuff.txt"), []byte("more stuff"), 0644)).To(Succeed())
+			Expect(os.WriteFile(filepath.Join(ctx.Application.Path, "target", "more-stuff-quotes.txt"), []byte(`before -jar "more stuff.jar" after -other="my path"`), 0644)).To(Succeed())
+			Expect(os.WriteFile(filepath.Join(ctx.Application.Path, "target", "more-stuff-class.txt"), []byte(`stuff -jar stuff.jar after`), 0644)).To(Succeed())
+			Expect(os.WriteFile(filepath.Join(ctx.Application.Path, "target", "override.txt"), []byte(`one=output`), 0644)).To(Succeed())
 		})
 
 		it("has none", func() {
@@ -146,7 +140,7 @@ func testArguments(t *testing.T, context spec.G, it spec.S) {
 			Expect(err).ToNot(HaveOccurred())
 			Expect(startClass).To(Equal(""))
 			Expect(args).To(HaveLen(4))
-			Expect(args).To(Equal([]string{"one", "two", "three", fmt.Sprintf("@%s",filepath.Join(ctx.Application.Path,"target/more-stuff.txt"))}))
+			Expect(args).To(Equal([]string{"one", "two", "three", fmt.Sprintf("@%s", filepath.Join(ctx.Application.Path, "target/more-stuff.txt"))}))
 		})
 
 		it("works with quotes in the file", func() {
@@ -158,7 +152,7 @@ func testArguments(t *testing.T, context spec.G, it spec.S) {
 			Expect(startClass).To(Equal(""))
 			Expect(args).To(HaveLen(4))
 			Expect(args).To(Equal([]string{"one", "two", "three", fmt.Sprintf("@%s", filepath.Join(ctx.Application.Path, "target/more-stuff-quotes.txt"))}))
-			bits, err := ioutil.ReadFile(filepath.Join(ctx.Application.Path, "target/more-stuff-quotes.txt"))
+			bits, err := os.ReadFile(filepath.Join(ctx.Application.Path, "target/more-stuff-quotes.txt"))
 			Expect(err).ToNot(HaveOccurred())
 			Expect(string(bits)).To(Equal("before after -other=\"my path\""))
 		})
@@ -170,9 +164,9 @@ func testArguments(t *testing.T, context spec.G, it spec.S) {
 			Expect(err).ToNot(HaveOccurred())
 			Expect(args).To(HaveLen(1))
 			Expect(args).To(Equal([]string{
-				fmt.Sprintf("@%s",filepath.Join(ctx.Application.Path, "target", "more-stuff-class.txt")),
+				fmt.Sprintf("@%s", filepath.Join(ctx.Application.Path, "target", "more-stuff-class.txt")),
 			}))
-			bits, err := ioutil.ReadFile(filepath.Join(ctx.Application.Path, "target/more-stuff-class.txt"))
+			bits, err := os.ReadFile(filepath.Join(ctx.Application.Path, "target/more-stuff-class.txt"))
 			Expect(err).ToNot(HaveOccurred())
 			Expect(string(bits)).To(Equal("after"))
 		})
@@ -254,9 +248,9 @@ func testArguments(t *testing.T, context spec.G, it spec.S) {
 	context("jar file", func() {
 		it.Before(func() {
 			Expect(os.MkdirAll(filepath.Join(ctx.Application.Path, "target"), 0755)).To(Succeed())
-			Expect(ioutil.WriteFile(filepath.Join(ctx.Application.Path, "target", "found.jar"), []byte{}, 0644)).To(Succeed())
-			Expect(ioutil.WriteFile(filepath.Join(ctx.Application.Path, "target", "a.two"), []byte{}, 0644)).To(Succeed())
-			Expect(ioutil.WriteFile(filepath.Join(ctx.Application.Path, "target", "b.two"), []byte{}, 0644)).To(Succeed())
+			Expect(os.WriteFile(filepath.Join(ctx.Application.Path, "target", "found.jar"), []byte{}, 0644)).To(Succeed())
+			Expect(os.WriteFile(filepath.Join(ctx.Application.Path, "target", "a.two"), []byte{}, 0644)).To(Succeed())
+			Expect(os.WriteFile(filepath.Join(ctx.Application.Path, "target", "b.two"), []byte{}, 0644)).To(Succeed())
 		})
 
 		it("adds arguments", func() {

--- a/native/build.go
+++ b/native/build.go
@@ -19,9 +19,10 @@ package native
 import (
 	"errors"
 	"fmt"
-	"github.com/paketo-buildpacks/libpak/sherpa"
 	"os"
 	"path/filepath"
+
+	"github.com/paketo-buildpacks/libpak/sherpa"
 
 	"github.com/paketo-buildpacks/libpak/effect"
 	"github.com/paketo-buildpacks/libpak/sbom"

--- a/native/native_image.go
+++ b/native/native_image.go
@@ -20,12 +20,11 @@ import (
 	"bytes"
 	"crypto/sha256"
 	"fmt"
-	"github.com/paketo-buildpacks/native-image/v5/native/slices"
-	"io"
-	"io/ioutil"
 	"os"
 	"path/filepath"
 	"strings"
+
+	"github.com/paketo-buildpacks/native-image/v5/native/slices"
 
 	"github.com/buildpacks/libcnb"
 	"github.com/magiconair/properties"
@@ -143,7 +142,7 @@ func (n NativeImage) Contribute(layer libcnb.Layer) (libcnb.Layer, error) {
 	}
 
 	n.Logger.Header("Removing bytecode")
-	cs, err := ioutil.ReadDir(n.ApplicationPath)
+	cs, err := os.ReadDir(n.ApplicationPath)
 	if err != nil {
 		return libcnb.Layer{}, fmt.Errorf("unable to list children of %s\n%w", n.ApplicationPath, err)
 	}
@@ -154,22 +153,8 @@ func (n NativeImage) Contribute(layer libcnb.Layer) (libcnb.Layer, error) {
 		}
 	}
 
-	src := filepath.Join(layer.Path, startClass)
-	in, err := os.Open(src)
-	if err != nil {
-		return libcnb.Layer{}, fmt.Errorf("unable to open %s\n%w", filepath.Join(layer.Path, startClass), err)
-	}
-	defer in.Close()
-
-	dst := filepath.Join(n.ApplicationPath, startClass)
-	out, err := os.OpenFile(dst, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0755)
-	if err != nil {
-		return libcnb.Layer{}, fmt.Errorf("unable to open %s\n%w", dst, err)
-	}
-	defer out.Close()
-
-	if _, err := io.Copy(out, in); err != nil {
-		return libcnb.Layer{}, fmt.Errorf("unable to copy\n%w", err)
+	if err := copyFilesFromLayer(layer.Path, startClass, n.ApplicationPath); err != nil {
+		return libcnb.Layer{}, fmt.Errorf("unable to copy files from layer\n%w", err)
 	}
 
 	return layer, nil
@@ -224,4 +209,43 @@ func (n NativeImage) ProcessArguments(layer libcnb.Layer) ([]string, string, err
 
 func (NativeImage) Name() string {
 	return "native-image"
+}
+
+// copy the main file & any `*.so` files also in the layer to the application path
+func copyFilesFromLayer(layerPath string, execName string, appPath string) error {
+	files, err := os.ReadDir(layerPath)
+	if err != nil {
+		return fmt.Errorf("unable to list files on layer %s\n%w", layerPath, err)
+	}
+
+	for _, file := range files {
+		if file.Type().IsRegular() && (file.Name() == execName) {
+			src := filepath.Join(layerPath, file.Name())
+			dst := filepath.Join(appPath, file.Name())
+
+			if err := copyFile(src, dst); err != nil {
+				return fmt.Errorf("unable to copy %s to %s\n%w", src, dst, err)
+			}
+		}
+		if file.Type().IsRegular() && (strings.HasSuffix(file.Name(), ".so")) {
+			src := filepath.Join(layerPath, file.Name())
+			dst := filepath.Join(appPath, file.Name())
+
+			if err := copyFile(src, dst); err != nil {
+				return fmt.Errorf("unable to copy %s to %s\n%w", src, dst, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func copyFile(src string, dst string) error {
+	in, err := os.Open(src)
+	if err != nil {
+		return fmt.Errorf("unable to open %s\n%w", src, err)
+	}
+	defer in.Close()
+
+	return sherpa.CopyFile(in, dst)
 }

--- a/native/native_image_test.go
+++ b/native/native_image_test.go
@@ -19,7 +19,6 @@ package native_test
 import (
 	"fmt"
 	"io"
-	"io/ioutil"
 	"os"
 	"path/filepath"
 	"strings"
@@ -50,27 +49,22 @@ func testNativeImage(t *testing.T, context spec.G, it spec.S) {
 	)
 
 	it.Before(func() {
-		var err error
-
-		ctx.Application.Path, err = ioutil.TempDir("", "native-image-application")
-		Expect(err).NotTo(HaveOccurred())
-
-		ctx.Layers.Path, err = ioutil.TempDir("", "native-image-layers")
-		Expect(err).NotTo(HaveOccurred())
+		ctx.Application.Path = t.TempDir()
+		ctx.Layers.Path = t.TempDir()
 
 		executor = &mocks.Executor{}
 
 		props = properties.NewProperties()
 
-		_, _, err = props.Set("Start-Class", "test-start-class")
+		_, _, err := props.Set("Start-Class", "test-start-class")
 		Expect(err).NotTo(HaveOccurred())
 		_, _, err = props.Set("Class-Path", "manifest-class-path")
 		Expect(err).NotTo(HaveOccurred())
 
-		Expect(ioutil.WriteFile(filepath.Join(ctx.Application.Path, "fixture-marker"), []byte{}, 0644)).To(Succeed())
+		Expect(os.WriteFile(filepath.Join(ctx.Application.Path, "fixture-marker"), []byte{}, 0644)).To(Succeed())
 		Expect(os.MkdirAll(filepath.Join(ctx.Application.Path, "BOOT-INF"), 0755)).To(Succeed())
 		Expect(os.MkdirAll(filepath.Join(ctx.Application.Path, "META-INF"), 0755)).To(Succeed())
-		Expect(ioutil.WriteFile(filepath.Join(ctx.Application.Path, "META-INF", "MANIFEST.MF"), []byte{}, 0644)).To(Succeed())
+		Expect(os.WriteFile(filepath.Join(ctx.Application.Path, "META-INF", "MANIFEST.MF"), []byte{}, 0644)).To(Succeed())
 
 		nativeImage, err = native.NewNativeImage(ctx.Application.Path, "test-argument-1 test-argument-2", "", "none", "", props, ctx.StackID)
 		nativeImage.Logger = bard.NewLogger(io.Discard)
@@ -91,7 +85,9 @@ func testNativeImage(t *testing.T, context spec.G, it spec.S) {
 		})).Run(func(args mock.Arguments) {
 			exec := args.Get(0).(effect.Execution)
 			lastArg := exec.Args[len(exec.Args)-1]
-			Expect(ioutil.WriteFile(filepath.Join(layer.Path, lastArg), []byte{}, 0644)).To(Succeed())
+			Expect(os.WriteFile(filepath.Join(layer.Path, lastArg), []byte{}, 0755)).To(Succeed())
+			Expect(os.WriteFile(filepath.Join(layer.Path, "libawt.so"), []byte{}, 0644)).To(Succeed())
+			Expect(os.WriteFile(filepath.Join(layer.Path, "libawt_headless.so"), []byte{}, 0644)).To(Succeed())
 		}).Return(nil)
 
 		executor.On("Execute", mock.MatchedBy(func(e effect.Execution) bool {
@@ -100,7 +96,9 @@ func testNativeImage(t *testing.T, context spec.G, it spec.S) {
 		})).Run(func(args mock.Arguments) {
 			exec := args.Get(0).(effect.Execution)
 			lastArg := exec.Args[len(exec.Args)-1]
-			Expect(ioutil.WriteFile(filepath.Join(layer.Path, lastArg), []byte{}, 0644)).To(Succeed())
+			Expect(os.WriteFile(filepath.Join(layer.Path, lastArg), []byte{}, 0755)).To(Succeed())
+			Expect(os.WriteFile(filepath.Join(layer.Path, "libawt.so"), []byte{}, 0644)).To(Succeed())
+			Expect(os.WriteFile(filepath.Join(layer.Path, "libawt_headless.so"), []byte{}, 0644)).To(Succeed())
 		}).Return(nil)
 
 		layer, err = ctx.Layers.Layer("test-layer")
@@ -134,6 +132,25 @@ func testNativeImage(t *testing.T, context spec.G, it spec.S) {
 				"-cp", "some-classpath",
 				"test-start-class",
 			}))
+
+			Expect(filepath.Join(ctx.Application.Path, "BOOT-INF")).ToNot(BeADirectory())
+			Expect(filepath.Join(ctx.Application.Path, "META-INF")).ToNot(BeADirectory())
+
+			Expect(filepath.Join(layer.Path, "test-start-class")).To(BeARegularFile())
+			Expect(filepath.Join(layer.Path, "libawt.so")).To(BeARegularFile())
+			Expect(filepath.Join(layer.Path, "libawt_headless.so")).To(BeARegularFile())
+
+			info, err := os.Stat(filepath.Join(layer.Path, "test-start-class"))
+			Expect(err).NotTo(HaveOccurred())
+			fmt.Println("info.Mode().Perm(): ", info.Mode().Perm().String())
+			Expect(info.Mode().Perm()).To(Equal(os.FileMode(0755)))
+
+			Expect(filepath.Join(ctx.Application.Path, "test-start-class")).To(BeARegularFile())
+			Expect(filepath.Join(ctx.Application.Path, "libawt.so")).To(BeARegularFile())
+			Expect(filepath.Join(ctx.Application.Path, "libawt_headless.so")).To(BeARegularFile())
+			info, err = os.Stat(filepath.Join(ctx.Application.Path, "test-start-class"))
+			Expect(err).NotTo(HaveOccurred())
+			Expect(info.Mode().Perm()).To(Equal(os.FileMode(0755)))
 		})
 	})
 
@@ -160,7 +177,7 @@ func testNativeImage(t *testing.T, context spec.G, it spec.S) {
 		it("contributes native image with Class-Path from manifest and args from a file", func() {
 			argsFile := filepath.Join(ctx.Application.Path, "target", "args.txt")
 			Expect(os.MkdirAll(filepath.Join(ctx.Application.Path, "target"), 0755)).To(Succeed())
-			Expect(ioutil.WriteFile(argsFile, []byte(`test-argument-1 test-argument-2`), 0644)).To(Succeed())
+			Expect(os.WriteFile(argsFile, []byte(`test-argument-1 test-argument-2`), 0644)).To(Succeed())
 
 			nativeImage, err := native.NewNativeImage(ctx.Application.Path, "", argsFile, "none", "", props, ctx.StackID)
 			nativeImage.Logger = bard.NewLogger(io.Discard)
@@ -209,7 +226,7 @@ func testNativeImage(t *testing.T, context spec.G, it spec.S) {
 			})).Run(func(args mock.Arguments) {
 				exec := args.Get(0).(effect.Execution)
 				lastArg := exec.Args[len(exec.Args)-1]
-				Expect(ioutil.WriteFile(filepath.Join(layer.Path, lastArg), []byte{}, 0644)).To(Succeed())
+				Expect(os.WriteFile(filepath.Join(layer.Path, lastArg), []byte{}, 0644)).To(Succeed())
 			}).Return(nil)
 
 			layer, err = ctx.Layers.Layer("test-layer")
@@ -254,7 +271,7 @@ func testNativeImage(t *testing.T, context spec.G, it spec.S) {
 			})).Run(func(args mock.Arguments) {
 				exec := args.Get(0).(effect.Execution)
 				lastArg := exec.Args[len(exec.Args)-1]
-				Expect(ioutil.WriteFile(filepath.Join(layer.Path, lastArg), []byte{}, 0644)).To(Succeed())
+				Expect(os.WriteFile(filepath.Join(layer.Path, lastArg), []byte{}, 0644)).To(Succeed())
 			}).Return(nil)
 
 			layer, err = ctx.Layers.Layer("test-layer")
@@ -316,7 +333,7 @@ func testNativeImage(t *testing.T, context spec.G, it spec.S) {
 			executor.On("Execute", mock.MatchedBy(func(e effect.Execution) bool {
 				return e.Command == "upx"
 			})).Run(func(args mock.Arguments) {
-				Expect(ioutil.WriteFile(filepath.Join(layer.Path, "test-start-class"), []byte("upx-compressed"), 0644)).To(Succeed())
+				Expect(os.WriteFile(filepath.Join(layer.Path, "test-start-class"), []byte("upx-compressed"), 0644)).To(Succeed())
 			}).Return(nil)
 
 			_, err := nativeImage.Contribute(layer)
@@ -331,7 +348,7 @@ func testNativeImage(t *testing.T, context spec.G, it spec.S) {
 			bin := filepath.Join(layer.Path, "test-start-class")
 			Expect(bin).To(BeARegularFile())
 
-			data, err := ioutil.ReadFile(bin)
+			data, err := os.ReadFile(bin)
 			Expect(err).ToNot(HaveOccurred())
 			Expect(data).To(ContainSubstring("upx-compressed"))
 		})
@@ -344,8 +361,8 @@ func testNativeImage(t *testing.T, context spec.G, it spec.S) {
 			executor.On("Execute", mock.MatchedBy(func(e effect.Execution) bool {
 				return e.Command == "gzexe"
 			})).Run(func(args mock.Arguments) {
-				Expect(ioutil.WriteFile(filepath.Join(layer.Path, "test-start-class"), []byte("gzexe-compressed"), 0644)).To(Succeed())
-				Expect(ioutil.WriteFile(filepath.Join(layer.Path, "test-start-class~"), []byte("original"), 0644)).To(Succeed())
+				Expect(os.WriteFile(filepath.Join(layer.Path, "test-start-class"), []byte("gzexe-compressed"), 0644)).To(Succeed())
+				Expect(os.WriteFile(filepath.Join(layer.Path, "test-start-class~"), []byte("original"), 0644)).To(Succeed())
 			}).Return(nil)
 
 			_, err := nativeImage.Contribute(layer)
@@ -360,7 +377,7 @@ func testNativeImage(t *testing.T, context spec.G, it spec.S) {
 			bin := filepath.Join(layer.Path, "test-start-class")
 			Expect(bin).To(BeARegularFile())
 
-			data, err := ioutil.ReadFile(bin)
+			data, err := os.ReadFile(bin)
 			Expect(err).ToNot(HaveOccurred())
 			Expect(data).To(ContainSubstring("gzexe-compressed"))
 			Expect(filepath.Join(layer.Path, "test-start-class~")).ToNot(BeAnExistingFile())
@@ -373,7 +390,7 @@ func testNativeImage(t *testing.T, context spec.G, it spec.S) {
 		})
 
 		it("contributes a static native image executable with dynamic libc", func() {
-			Expect(ioutil.WriteFile(filepath.Join(ctx.Application.Path, "BOOT-INF", "classpath.idx"), []byte(`
+			Expect(os.WriteFile(filepath.Join(ctx.Application.Path, "BOOT-INF", "classpath.idx"), []byte(`
 - "test-jar.jar"
 - "spring-graalvm-native-0.8.6-xxxxxx.jar"
 `), 0644)).To(Succeed())

